homeassistant/components/kodi/binary_sensor.py

```diff
@@ -0,0 +1,166 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DATA_CONNECTION, DOMAIN
+from .kodi_connman import KodiConnectionClient, KodiConnectionManager
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Kodi sensors based on a config entry."""
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    connman = data[DATA_CONNECTION]
+    if (uid := config_entry.unique_id) is None:
+        uid = config_entry.entry_id
+
+    sensor_entities = [
+        KodiBinaryEntity(
+            "Screensaver",
+            "GUI.OnScreensaverActivated",
+            "GUI.OnScreensaverDeactivated",
+            "System.ScreenSaverActive",
+            uid,
+            connman,
+        ),
+        KodiBinaryEntity(
+            "Energy saving",
+            "GUI.OnDPMSActivated",
+            "GUI.OnDPMSDeactivated",
+            "System.DPMSActive",
+            uid,
+            connman,
+        ),
+    ]
+    async_add_entities(sensor_entities)
+
+
+class KodiBinaryEntity(KodiConnectionClient, BinarySensorEntity):
+    """Generic binary sensor entity for WebSocket callbacks."""
+
+    def __init__(
+        self,
+        name: str,
+        api_on: str,
+        api_off: str,
+        boolean_name: str,
+        uid: str,
+        connman: KodiConnectionManager,
+    ) -> None:
+        """Initialize kodi binary sensor."""
+        super().__init__(connman)
+        self._attr_is_on = False
+        self._attr_name = name
+        self._attr_unique_id = uid + "_" + api_on
+        self._api_on = api_on
+        self._api_off = api_off
+        self._kodi_boolean = boolean_name
+        self._kodi_uid = uid
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self._kodi_uid)},
+        )
+
+    @callback
+    def async_on(self, sender, data):  # pylint: disable=unused-argument
+        """Switch sensor on from api call."""
+        self._attr_is_on = True
+        self.async_write_ha_state()
+        _LOGGER.debug("Kodi %s on (%s)", self.name, self._api_on)
+
+    @callback
+    def async_off(self, sender, data):  # pylint: disable=unused-argument
+        """Switch sensor off from api call."""
+        self._attr_is_on = False
+        self.async_write_ha_state()
+        _LOGGER.debug("Kodi %s off (%s)", self.name, self._api_off)
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks for needed api endpoints."""
+        await self._connman.add_callback_on_connect(self._on_ws_connected)
+        await self._connman.add_callback_on_disconnect(self._on_ws_disconnected)
+
+    @callback
+    async def _on_ws_connected(self):
+        """Call after websocket is connected."""
+        _LOGGER.debug("Binary sensor %s websocket connected", self.name)
+        await self._async_query_booleans()
```

Comment:
can this be replaced with `self. async_schedule_update_ha_state(True)`?
That way you can push it to the base class, and get rid of `_on_ws_connected` (and same can probably be done for `_on_ws_disconnected`

Reply:
Yes indeed. Done.